### ISSUE TITLE
Monitoring static pods in k8s

### ISSUE THREAD

**Issue by andor44** [DataDog/datadog-agent#2803]
This is essentially a follow-up to an issue we had before, namely #1447

Long story short, the kubelet will always show [static pods](https://kubernetes.io/docs/tasks/administer-cluster/static-pod/) as status `Pending`. This was causing issues for the agent as seen on the ticket above, so as far as I can tell the decision was made to make the agent ignore `Pending` pods altogether.

However, we use static pods to run k8s master components (API server, scheduler, etc.) that we'd like to monitor with Datadog through autodiscovery annotations. Right now the annotations are completely ignored, I assume because all `Pending` pods are ignored.

Of course in an ideal scenario this bug would be fixed upstream so that even static pods are correctly reported as `Running` the kubelet's local pod list but that issue has been standing still for more than 6 months. Would you consider special-casing static pods so even if they're `Pending` checks will be ran against them? You can tell static pods apart from regular pods because the kubelet automatically puts a `"kubernetes.io/config.source": "file"` annotation on them.

**Comment by rifelpet** [comment]
Yes please! We specifically run into this issue on our kubernetes clusters provisioned with [Kops](https://github.com/kubernetes/kops). Kops launches etcd, kube-controller-manager, kube-apiserver, and kube-proxy as static pods. This means that we cant use autodiscovery for etcd or kube-proxy. I've talked to Datadog support a few times about this and the recommendation is to run "dummy pods" on the same nodes as the static pods:

```yaml
      containers:
      - name: dummy
        image: k8s.gcr.io/etcd:3.2.18
        command: ["/bin/sh"]
        args:
        - -c
        - while :; do sleep 2073600; done
```

This way the agent will discover that etcd is running on the same node and perform the checks. It requires hardcoding the hostname and port in the check definition so that the agent doesn't try to reach the dummy pod's IP and instead connects to the real etcd.

I believe the fix on the agent side would either be [here](https://github.com/DataDog/datadog-agent/blob/0e8c3655399e85740098b23d931bf835009065c7/pkg/autodiscovery/listeners/kubelet.go#L120) or [here](https://github.com/DataDog/datadog-agent/blob/eaf05676b3e317c5bbea37c4b12341f8446c5d3d/pkg/util/kubernetes/kubelet/kubelet.go#L539).

Please let me get rid of all these dummy pods!

**Comment by hkaj** [comment]
#2970 makes the podwatcher aware of static pods, but we won't be able to enable autodiscovery on them because the pod IP is part of the status that's not updated in the pod list.

There are several ways forward:

* we can add static pods in autodiscovery, but you won't be able to use `%%host%%` in config templates as we won't can't get the pod IP. It can still be useful if your pod has a service name/domain name you can hardcode in the URL, if it exposes a hostPort (and use `hostIP:hostPort` in the config template), or if everything runs with `hostNetwork:true` in which case you can hardcode localhost in the template
* we watch the pod list from the apiserver, except it's not scalable to do it from every agent, so we'll have to do it in the cluster agent and proxy the info to agents. This would add a dependency for this feature on the cluster agent, and would also make it use *a lot* of RAM
* we try revisiting fixing the pod list upstream

If the first solution still sounds useful we can implement that quickly, but I don't think it's very valuable.

Curious to hear your thoughts on this @andor44 @rifelpet

**Comment by hkaj** [comment]
Just posted on https://github.com/kubernetes/kubernetes/issues/61717 if you want to track solution 3

**Comment by andor44** [comment]
Ha, good catch on that one.

What I can say is that in our case (so I guess in @rifelpet's too) solution 1 works as we're using static pods to run master components and kube-proxy, all of which need to be `hostNetwork: true` by their nature. I'd be happy with that, that already fixes our immediate problem. Though that's also side-stepping the issue a bit.

I'm also fine with option two, though for example there's also #1853, which is also trying to shift heavy-lifting to DCA. Can the DCA take all this extra load? Is it worth adding all this extra complexity for an admittedly somewhat niche use-case? 🤷‍♂️ 

If you guys can push a fix upstream that'd be the bees knees 😄

**Comment by rifelpet** [comment]
Agreed, all of Kops' default static pods use host networking:
* [etcd](https://github.com/kubernetes/kops/blob/ea420dac78ba0f8f669a94984598ba4686400e57/protokube/pkg/protokube/etcd_manifest.go)
* [kube-proxy](https://github.com/kubernetes/kops/blob/ea420dac78ba0f8f669a94984598ba4686400e57/nodeup/pkg/model/kube_proxy.go#L203)
* [kube-scheduler](https://github.com/kubernetes/kops/blob/ea420dac78ba0f8f669a94984598ba4686400e57/nodeup/pkg/model/kube_scheduler.go#L121)
* [kube-controller-manager](https://github.com/kubernetes/kops/blob/ea420dac78ba0f8f669a94984598ba4686400e57/nodeup/pkg/model/kube_controller_manager.go#L150)

Additionally, the current workaround of using dummy pods doesn't support `%%host%%` since it would use the dummy pod's IP rather than the real pod's IP.

I'm fine with the `hostNetwork: true` requirement but obviously if we can get this fixed upstream that would be ideal.

**Comment by hkaj** [comment]
Update: we're working on the upstream solution, 6.10 doesn't make autodiscovery work with static pods. We'll keep this issue open for tracking.

**Comment by mfpierre** [comment]
Hey,

good new everyone 🎉  the fix for the static pods update on the kublet pod list has been merged https://github.com/kubernetes/kubernetes/pull/77661 and should be shipped with kubernetes `1.15`

We have opened cherry-pick PRs for kubernetes 1.12, 1.13, 1.14 but they're not merged yet 🤞
https://github.com/kubernetes/kubernetes/pull/77939
https://github.com/kubernetes/kubernetes/pull/77934
https://github.com/kubernetes/kubernetes/pull/77941

We have effectively tested this on a manually setup cluster with a modified kubelet and were able  monitor static pods with some AD annotations 👍

**Comment by grosser** [comment]
we are on kubernetes 1.17 + datadog 7.19.1 and this seems to not be working
the pod on the node with the static pod does not seem to be found by the datadog pod on that node
`agent status` shows other pods with annotations but not the static  :(

**Comment by jacjtito** [comment]
Hi @andor44 👋 — thanks again for taking the time to report this issue.

We're closing this due to inactivity: it's been over 6 months since the last update, and the stale bot's check-in didn't receive a response. This isn't a reflection of the value of the contribution — it's purely a hygiene pass to keep the backlog actionable.

If this is still relevant to you:
- Feel free to **reopen this issue** or open a new one with updated reproduction steps
- If the underlying issue is still present, a fresh report against the current version will get faster attention

Thanks for your patience and for contributing to the Datadog Agent! 🙏
